lib/plugin.js

```diff
@@ -50,106 +46,117 @@ class SVGSpritePlugin {
     return NAMESPACE;
   }
 
-  getReplacements(symbols, compilation) {
+  getReplacements() {
     const isPlainSprite = this.config.plainSprite === true;
-    const map = new MappedList(symbols, compilation);
-
     // TODO clarify what the hell `SymbolFile` means here
-    const replacements = map.groupItemsBySymbolFile((acc, item) => {
+    const replacements = this.map.groupItemsBySymbolFile((acc, item) => {
       acc[item.resource] = isPlainSprite ? item.url : item.useUrl;
     });
     return replacements;
   }
 
+  replaceSpriteInModuleSource(compilation) {
+    const { symbols } = this.svgCompiler;
+    this.map = new MappedList(symbols, compilation);
+    this.replacements = this.getReplacements(symbols, compilation, 1);
+    this.map.items.forEach(item => replaceInModuleSource(item.module, this.replacements));
+  }
+
+  replaceInModuleSource(compilation, chunks) {
+    chunks.forEach((chunk) => {
+      const modules = chunk.modules || Array.from(chunk.modulesIterable);
+
+      modules.forEach(() => {
+        if (Object.hasOwnProperty.call(module, '_originalModule')) {
+          replaceInModuleSource(module, this.replacements);
+        }
+      });
+    });
+  }
+
   // TODO optimize MappedList instantiation in each hook
   apply(compiler) {
-    const plugin = this;
+    const useNewTapableAPI = !!compiler.hooks;
+
     const { symbols } = this.svgCompiler;
 
-    // Handle only main compilation
-    compiler.plugin('this-compilation', (compilation) => {
-      // Share svgCompiler with loader
-      compilation.plugin('normal-module-loader', (loaderContext) => {
-        loaderContext[NAMESPACE] = plugin;
-      });
+    if (useNewTapableAPI) {
+      compiler.hooks
+      .thisCompilation
+      .tap(NAMESPACE, (compilation) => {
+        compilation.hooks
+          .normalModuleLoader
+          .tap(NAMESPACE, loaderContext => loaderContext[NAMESPACE] = this);
+
+        compilation.hooks.afterOptimizeChunks
+          .tap(NAMESPACE, () => this.replaceSpriteInModuleSource(compilation));
+
+        compilation.hooks.optimizeExtractedChunks
+          .tap(NAMESPACE, chunks => this.replaceInModuleSource(compilation, chunks));
+
+        if (compilation.hooks.htmlWebpackPluginBeforeHtmlGeneration) {
+          compilation.hooks
+            .htmlWebpackPluginBeforeHtmlGeneration
+            .tapAsync(NAMESPACE, (htmlPluginData, callback) => {
+              htmlPluginData.assets.sprites = generateSprites(this, compilation, symbols);
```

Comment:
I do think that the two `htmlWebpackPlugin` hooks don't need to be async though, I was just testing something out.

Reply:
Have no clue about that tbh, but expect some input from @kisenka anytime soon. (I hope)